dcrec/secp256k1/signature.go

```diff
@@ -56,36 +56,84 @@ func NewSignature(r, s *big.Int) *Signature {
 	return &Signature{r, s}
 }
 
-// Serialize returns the ECDSA signature in the more strict DER format.  Note
-// that the serialized bytes returned do not include the appended hash type used
-// in Decred signature scripts.
+// Serialize returns the ECDSA signature in the Distinguished Encoding Rules
+// (DER) format per section 10 of [ISO/IEC 8825-1] and such that the S component
+// of the signature is less than or equal to the half order of the group.
 //
-// 0x30 <length> 0x02 <length r> r 0x02 <length s> s
+// Note that the serialized bytes returned do not include the appended hash type
+// used in Decred signature scripts.
 func (sig *Signature) Serialize() []byte {
-	// Low 'S' malleability breaker.
-	sigS := sig.s
-	if sigS.Cmp(halforder) == 1 {
-		sigS = new(big.Int).Sub(order, sigS)
-	}
+	// The format of a DER encoded signature is as follows:
+	//
+	// 0x30 <total length> 0x02 <length of R> <R> 0x02 <length of S> <S>
+	//   - 0x30 is the ASN.1 identifier for a sequence
+	//   - Total length is 1 byte and specifies length of all remaining data
+	//   - 0x02 is the ASN.1 identifier that specifies an integer follows
+	//   - Length of R is 1 byte and specifies how many bytes R occupies
+	//   - R is the arbitrary length big-endian encoded number which
+	//     represents the R value of the signature.  DER encoding dictates
+	//     that the value must be encoded using the minimum possible number
+	//     of bytes.  This implies the first byte can only be null if the
+	//     highest bit of the next byte is set in order to prevent it from
+	//     being interpreted as a negative number.
+	//   - 0x02 is once again the ASN.1 integer identifier
+	//   - Length of S is 1 byte and specifies how many bytes S occupies
+	//   - S is the arbitrary length big-endian encoded number which
+	//     represents the S value of the signature.  The encoding rules are
+	//     identical as those for R.
+	const (
+		asn1SequenceID = 0x30
+		asn1IntegerID  = 0x02
+	)
 
-	// Ensure the encoded bytes for the R and S values are canonical and thus
-	// suitable for DER encoding.
-	rb := canonicalizeInt(sig.r)
-	sb := canonicalizeInt(sigS)
+	// Convert big ints to mod N scalars.  Ultimately the goal is to convert
+	// the signature type itself to use mod N scalars directly which will allow
+	// this step to be removed.
+	var r, s ModNScalar
+	r.SetByteSlice(sig.r.Bytes())
+	s.SetByteSlice(sig.s.Bytes())
+
+	// Ensure the S component of the signature is less than or equal to the half
+	// order of the group because both S and its negation are valid signatures
+	// modulo the order, so this forces a consistent choice to reduce signature
+	// malleability.
+	sigS := new(ModNScalar).Set(&s)
```

Comment:
is there a reason a new mod N scalar is created here for this negation, instead of negating the s directly?

Reply:
It was to minimize changes later when the signature itself uses mod N scalars and negating s directly would mutate the signature.